public/app/features/dashboard-scene/scene/AlertStatesDataLayer.ts

```diff
@@ -0,0 +1,227 @@
+
+import { getDashboardSceneFor } from '../utils/utils';
+
+interface AlertStatesDataLayerState extends SceneDataLayerProviderState {}
+
+export class AlertStatesDataLayer
+  extends SceneDataLayerBase<AlertStatesDataLayerState>
+  implements SceneDataLayerProvider
+{
+  private hasAlertRules = true;
+  private _timeRangeSub: Unsubscribable | undefined;
+  public topic = DataTopic.AlertStates;
+
+  public constructor(initialState: AlertStatesDataLayerState) {
+    super({
+      isEnabled: true,
+      ...initialState,
+      isHidden: true,
+    });
+  }
+
+  public onEnable(): void {
+    const timeRange = sceneGraph.getTimeRange(this);
+
+    this._timeRangeSub = timeRange.subscribeToState(() => {
+      this.runWithTimeRange(timeRange);
+    });
+  }
+
+  public onDisable(): void {
+    this._timeRangeSub?.unsubscribe();
+  }
+
+  public runLayer() {
+    const timeRange = sceneGraph.getTimeRange(this);
+    this.runWithTimeRange(timeRange);
+  }
+
+  private async runWithTimeRange(timeRange: SceneTimeRangeLike) {
+    const dashboard = getDashboardSceneFor(this);
+    const { uid, id } = dashboard.state;
+
+    if (this.querySub) {
+      this.querySub.unsubscribe();
+    }
+
+    if (!this.canWork(timeRange)) {
+      return;
+    }
+
+    let alerStatesExecution: Observable<AlertStateInfo[]> | undefined;
+
+    if (this.isUsingLegacyAlerting()) {
+      alerStatesExecution = from(
+        getBackendSrv().get(
+          '/api/alerts/states-for-dashboard',
+          {
+            dashboardId: id,
+          },
+          `dashboard-query-runner-alert-states-${id}`
+        )
+      ).pipe(map((alertStates) => alertStates));
+    } else {
+      alerStatesExecution = from(
+        getBackendSrv().get(
+          '/api/prometheus/grafana/api/v1/rules',
+          {
+            dashboard_uid: uid!,
+          },
+          `dashboard-query-runner-unified-alert-states-${id}`
+        )
+      ).pipe(
+        map((result: PromRulesResponse) => {
+          if (result.status === 'success') {
+            this.hasAlertRules = false;
+            const panelIdToAlertState: Record<number, AlertStateInfo> = {};
+
+            result.data.groups.forEach((group) =>
+              group.rules.forEach((rule) => {
+                if (isAlertingRule(rule) && rule.annotations && rule.annotations[Annotation.panelID]) {
+                  this.hasAlertRules = true;
+                  const panelId = Number(rule.annotations[Annotation.panelID]);
+                  const state = promAlertStateToAlertState(rule.state);
+
+                  // there can be multiple alerts per panel, so we make sure we get the most severe state:
+                  // alerting > pending > ok
+                  if (!panelIdToAlertState[panelId]) {
+                    panelIdToAlertState[panelId] = {
+                      state,
+                      id: Object.keys(panelIdToAlertState).length,
+                      panelId,
+                      dashboardId: id!,
```

Comment:
🤔 shouldn't we use `uid` instead?

Reply:
This code is just a copy of what's currently implemented in UnifiedAlertStatesWorker.